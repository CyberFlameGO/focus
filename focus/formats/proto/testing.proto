syntax = "proto3";

package testing;

message Foo {
	uint32 bar = 1;
}
