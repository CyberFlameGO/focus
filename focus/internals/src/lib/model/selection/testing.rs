use std::{collections::HashSet, path::Path, sync::Arc};

use focus_testing::{init_logging, ScratchGitRepo};

use anyhow::Result;
use focus_util::app::App;
use maplit::hashset;
use tempfile::TempDir;

use super::*;

struct Fixture {
    #[allow(dead_code)]
    dir: TempDir,
    underlying: ScratchGitRepo,
    app: Arc<App>,
}

impl Fixture {
    fn new() -> Result<Self> {
        let dir = TempDir::new()?;
        let path = dir.path().join("dense");
        let branch = String::from("main");
        let repo =
            ScratchGitRepo::new_copied_fixture(Path::new("bazel_java_example"), &path, &branch)?;
        let app = Arc::new(App::new(false)?);
        Ok(Self {
            dir,
            underlying: repo,
            app,
        })
    }

    fn repo(&self) -> Result<Repo> {
        Repo::open(self.underlying.path(), self.app.clone())
    }
}

/// Extract project names from the selection
fn project_names(selection: &Selection) -> HashSet<String> {
    selection
        .projects
        .iter()
        .map(|project| project.name.clone())
        .collect()
}

#[test]
fn repo_with_no_selections_returns_mandatory_projects_in_computed_selection() -> Result<()> {
    init_logging();

    let fixture = Fixture::new()?;
    let repo = fixture.repo()?;

    let selections = repo.selections()?;
    let computed_selection = selections.computed_selection()?;
    assert_eq!(
        project_names(&computed_selection),
        hashset! {String::from("mandatory")}
    );
    assert!(computed_selection.targets.is_empty());

    Ok(())
}

#[test]
fn modifying_and_saving_the_selection() -> Result<()> {
    init_logging();

    let fixture = Fixture::new()?;
    let repo = fixture.repo()?;

    let project_name = String::from("team_banzai/project_a");
    let target = Target::try_from("bazel://library_b/...")?;

    {
        let mut selections = repo.selections()?;
        selections.process(&vec![
            Operation {
                action: OperationAction::Add,
                underlying: Underlying::Project(project_name.clone()),
            },
            Operation {
                action: OperationAction::Add,
                underlying: Underlying::Target(target.clone()),
            },
        ])?;
        selections.save()?;
        let computed_selection = selections.computed_selection()?;
        assert_eq!(
            project_names(&computed_selection),
            hashset! {String::from("mandatory"), project_name.clone()}
        );
        assert_eq!(computed_selection.targets, hashset! {target.clone()});
    }

    {
        // Ensure that after loading from disk in a new instance, the selection is the same.
        let mut selections = repo.selections()?;
        let computed_selection = selections.computed_selection()?;
        assert_eq!(
            project_names(&computed_selection),
            hashset! {String::from("mandatory"), project_name.clone()}
        );
        assert_eq!(computed_selection.targets, hashset! {target.clone()});

        // Remove the target
        selections.process(&vec![Operation {
            action: OperationAction::Remove,
            underlying: Underlying::Target(target),
        }])?;
        selections.save()?;
        let computed_selection = selections.computed_selection()?;
        assert_eq!(
            project_names(&computed_selection),
            hashset! {String::from("mandatory"), project_name}
        );
        assert!(computed_selection.targets.is_empty());
    }

    Ok(())
}

#[test]
fn adding_an_unknown_project() -> Result<()> {
    init_logging();

    let fixture = Fixture::new()?;
    let repo = fixture.repo()?;

    let nonexistent_project = Underlying::Project(String::from("blofeld/moonbase"));
    let mut selections = repo.selections()?;
    let result = selections.process(&vec![Operation {
        action: OperationAction::Add,
        underlying: nonexistent_project.clone(),
    }])?;
    assert_eq!(result.is_success(), false);
    assert_eq!(result.absent, hashset! {nonexistent_project});

    Ok(())
}

#[test]
fn mandatory_projects_cannot_be_selected() -> Result<()> {
    init_logging();

    let fixture = Fixture::new()?;
    let repo = fixture.repo()?;

    let mandatory_project = Underlying::Project(String::from("mandatory"));
    let mut selections = repo.selections()?;
    let result = selections.process(&vec![Operation {
        action: OperationAction::Add,
        underlying: mandatory_project.clone(),
    }])?;
    assert_eq!(result.is_success(), false);
    assert_eq!(result.absent, hashset! {mandatory_project});

    Ok(())
}

#[test]
fn duplicate_projects_are_ignored() -> Result<()> {
    init_logging();

    let fixture = Fixture::new()?;
    let repo = fixture.repo()?;

    let project_b = Underlying::Project(String::from("team_zissou/project_b"));
    let mut selections = repo.selections()?;
    let result = selections.process(&vec![Operation {
        action: OperationAction::Add,
        underlying: project_b.clone(),
    }])?;
    assert_eq!(result.is_success(), true);
    assert_eq!(result.added, hashset! {project_b.clone()});

    let result = selections.process(&vec![Operation {
        action: OperationAction::Add,
        underlying: project_b.clone(),
    }])?;
    assert_eq!(result.is_success(), true);
    assert_eq!(result.ignored, hashset! {project_b});

    Ok(())
}
