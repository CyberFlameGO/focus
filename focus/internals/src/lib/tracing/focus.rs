// module to collect tracing setup and config for the focus app itself

use std::io;

use anyhow::Result;
use tracing::dispatcher::DefaultGuard;
use tracing_appender::non_blocking::WorkerGuard;
use tracing_error::ErrorLayer;
use tracing_subscriber::prelude::*;
use tracing_subscriber::{self, util::SubscriberInitExt, EnvFilter};

#[derive(Debug)]
pub enum GuardWrapper {
    WorkerGuard(WorkerGuard),
    DefaultGuard(DefaultGuard),
}

#[derive(Debug, Default)]
/// opaque struct for returning tracing WorkerGuard instances to main
pub struct Guard {
    _inner: Vec<GuardWrapper>,
}

impl From<WorkerGuard> for Guard {
    fn from(wg: WorkerGuard) -> Self {
        Guard {
            _inner: vec![GuardWrapper::WorkerGuard(wg)],
        }
    }
}

#[derive(Debug, Default)]
pub struct TracingOpts {
    pub is_tty: bool,
    pub nocolor_requested: bool,
}

pub fn init_tracing(opts: TracingOpts) -> Result<Guard> {
    let TracingOpts {
        is_tty,
        nocolor_requested,
    } = opts;
    let use_color = is_tty && !nocolor_requested;
    let (non_blocking, guard) = tracing_appender::non_blocking(io::stderr());

    tracing_subscriber::registry()
        .with(ErrorLayer::default())
        .with(EnvFilter::new(
            std::env::var("RUST_LOG").unwrap_or_else(|_| "info".to_string()),
        ))
        .with(
            tracing_subscriber::fmt::layer()
                .with_span_events(tracing_subscriber::fmt::format::FmtSpan::CLOSE)
                .with_target(false)
                .with_writer(non_blocking)
                .with_ansi(use_color),
        )
        .try_init()?;

    Ok(Guard::from(guard))
}
