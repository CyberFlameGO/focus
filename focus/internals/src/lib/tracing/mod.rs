pub mod chrome;
pub mod focus;
pub mod git_trace2;

use std::path::PathBuf;

pub use chrome::Trace;
pub use focus::{init_tracing, Guard, TracingOpts};

use anyhow::Result;

#[cfg(test)]
pub(in crate::tracing) mod testing {
    use std::path::{Path, PathBuf};

    use anyhow::{anyhow, Result};

    pub const FIXTURE_RELPATH: &str = "src/lib/tracing/fixtures";

    pub fn manifest_dir() -> Result<PathBuf> {
        Ok(PathBuf::from(
            std::env::var_os("CARGO_MANIFEST_DIR").ok_or(anyhow!("CARGO_MANIFEST_DIR not set"))?,
        ))
    }

    pub fn manifest_relative_path<S: AsRef<Path>>(s: S) -> Result<PathBuf> {
        Ok(manifest_dir()?.join(s.as_ref()))
    }

    pub fn fixture_path<S: AsRef<Path>>(s: S) -> Result<PathBuf> {
        manifest_relative_path(FIXTURE_RELPATH).map(|p| p.join(s.as_ref()))
    }
}

/// returns the default system specific log location
#[cfg(target_os = "macos")]
pub fn log_dir() -> Result<PathBuf> {
    match dirs::home_dir().map(|pb| pb.join("Library/Logs/focus")) {
        Some(path) => Ok(path),
        None => Err(anyhow::anyhow!("HOME not defined")),
    }
}

#[cfg(not(any(
    target_os = "windows",
    target_os = "macos",
    target_os = "ios",
    target_arch = "wasm32"
)))]
pub fn log_dir() -> Result<PathBuf> {
    dirs::home_dir()
        .ok_or(anyhow::anyhow!("HOME not defined"))
        .map(|pb| pb.join(".local/focus/log"))
}
