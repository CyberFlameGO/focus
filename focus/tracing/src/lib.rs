// Copyright 2022 Twitter, Inc.
// SPDX-License-Identifier: Apache-2.0

pub mod chrome;
pub mod focus;
pub mod git_trace2;

use std::path::{Path, PathBuf};

pub use crate::focus::{init_tracing, Guard, TracingOpts};
pub use chrome::Trace;

use anyhow::Result;

#[cfg(test)]
pub(crate) mod testing {
    use std::path::{Path, PathBuf};

    use anyhow::{anyhow, Result};

    pub const FIXTURE_RELPATH: &str = "src/fixtures";

    pub fn manifest_dir() -> Result<PathBuf> {
        Ok(PathBuf::from(
            std::env::var_os("CARGO_MANIFEST_DIR")
                .ok_or_else(|| anyhow!("CARGO_MANIFEST_DIR not set"))?,
        ))
    }

    pub fn manifest_relative_path<S: AsRef<Path>>(s: S) -> Result<PathBuf> {
        Ok(manifest_dir()?.join(s.as_ref()))
    }

    pub fn fixture_path<S: AsRef<Path>>(s: S) -> Result<PathBuf> {
        manifest_relative_path(FIXTURE_RELPATH).map(|p| p.join(s.as_ref()))
    }
}

fn home_relative_path<P: AsRef<Path>>(p: P) -> Result<PathBuf> {
    match dirs::home_dir().map(|pb| pb.join(p.as_ref())) {
        Some(path) => Ok(path),
        None => Err(anyhow::anyhow!("HOME not defined")),
    }
}

#[cfg(target_os = "macos")]
const DEFAULT_LOG_DIR: &str = "Library/Logs/focus";

#[cfg(not(any(
    target_os = "windows",
    target_os = "macos",
    target_os = "ios",
    target_arch = "wasm32"
)))]
const DEFAULT_LOG_DIR: &str = ".local/focus/log";

/// returns the default system specific log location
pub fn log_dir() -> Result<PathBuf> {
    home_relative_path(DEFAULT_LOG_DIR)
}
