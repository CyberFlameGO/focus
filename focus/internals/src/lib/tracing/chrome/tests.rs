#![cfg(test)]

use super::*;
use crate::tracing::testing::*;
use anyhow::Result;

fn assert_trace_snapshot<S: AsRef<str>>(input: S) -> Result<()> {
    let trace = Trace::git_trace_from_file(&fixture_path(input.as_ref())?)?;
    insta::assert_yaml_snapshot!(input.as_ref(), trace);
    Ok(())
}

#[test]
fn test_clone_focus_trace() -> Result<()> {
    assert_trace_snapshot("clone-perf.json")
}

#[test]
fn test_status_trace() -> Result<()> {
    assert_trace_snapshot("status.json")
}

#[test]
fn test_push_trace() -> Result<()> {
    assert_trace_snapshot("push.json")
}

#[test]
fn test_reset_trace() -> Result<()> {
    assert_trace_snapshot("reset.json")
}

#[test]
fn test_reflog_trace() -> Result<()> {
    assert_trace_snapshot("reflog.json")
}

#[test]
fn test_rebase_trace() -> Result<()> {
    assert_trace_snapshot("rebase.json")
}

#[test]
fn test_branch_update_trace() -> Result<()> {
    assert_trace_snapshot("branch-update.json")
}
