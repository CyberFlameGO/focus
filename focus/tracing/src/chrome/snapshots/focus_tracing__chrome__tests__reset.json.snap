---
source: focus/internals/src/lib/tracing/chrome/tests.rs
assertion_line: 12
expression: trace

---
traceEvents:
  - ts: 0
    pid: 1
    tid: 1
    name: "data(traverse_trees): statistics -> {\"traverse_trees_count\":1,\"traverse_trees_max_depth\":1}"
    cat: ""
    ph: X
    dur: 11805
    args:
      category: traverse_trees
      file_line_thread: "tree-walk.c:185:main"
      git_event_name: data_json
      key: statistics
      nesting: 1
      t_abs: 0.011805
      t_rel: 0.011805
      value:
        traverse_trees_count: 1
        traverse_trees_max_depth: 1
  - ts: 1271
    pid: 1
    tid: 0
    name: main
    cat: ""
    ph: X
    dur: 10546
    args:
      begin:
        file_line_thread: "common-main.c:49:main"
        git_event_name: version
        git_version: 2.35.1
        stats_version: "3"
      end:
        code: 0
        file_line_thread: "trace2/tr2_tgt_event.c:201:main"
        git_event_name: atexit
        t_abs: 0.011817
  - ts: 1333
    pid: 1
    tid: 0
    name: git reset --hard ccdd887
    cat: ""
    ph: X
    dur: 10453
    args:
      begin:
        argv:
          - git
          - reset
          - "--hard"
          - ccdd887
        file_line_thread: "common-main.c:50:main"
        git_event_name: start
        t_abs: 0.001331
      end:
        code: 0
        file_line_thread: "git.c:716:main"
        git_event_name: exit
        t_abs: 0.011785
  - ts: 1622
    pid: 1
    tid: 0
    name: "repository.c:136:main"
    cat: ""
    ph: i
    args:
      file_line_thread: "repository.c:136:main"
      git_event_name: def_repo
      worktree: /Users/jsimms/workspace/focus
  - ts: 1803
    pid: 1
    tid: 0
    name: "git.c:459:main"
    cat: ""
    ph: i
    args:
      file_line_thread: "git.c:459:main"
      git_event_name: cmd_name
      hierarchy: reset
      name: reset
  - ts: 3260
    pid: 1
    tid: 0
    name: "builtin/reset.c:483:main"
    cat: ""
    ph: i
    args:
      file_line_thread: "builtin/reset.c:483:main"
      git_event_name: cmd_mode
      name: hard
  - ts: 3337
    pid: 1
    tid: 1
    name: do_read_index
    cat: ""
    ph: X
    dur: 270
    args:
      begin:
        category: index
        file_line_thread: "read-cache.c:2416:main"
        git_event_name: region_enter
        label: do_read_index
        msg: ".git/index"
        nesting: 1
      end:
        file_line_thread: "read-cache.c:2421:main"
        git_event_name: region_leave
        t_rel: 0.00027
  - ts: 3338
    pid: 1
    tid: 2
    name: "data(index): read/version -> 2"
    cat: ""
    ph: X
    dur: 199
    args:
      category: index
      file_line_thread: "read-cache.c:2362:main"
      git_event_name: data
      key: read/version
      nesting: 2
      t_abs: 0.003536
      t_rel: 0.000199
      value: "2"
  - ts: 3359
    pid: 1
    tid: 2
    name: "data(index): read/cache_nr -> 94"
    cat: ""
    ph: X
    dur: 234
    args:
      category: index
      file_line_thread: "read-cache.c:2364:main"
      git_event_name: data
      key: read/cache_nr
      nesting: 2
      t_abs: 0.003571
      t_rel: 0.000234
      value: "94"
  - ts: 3465
    pid: 1
    tid: 2
    name: read
    cat: ""
    ph: X
    dur: 30
    args:
      begin:
        category: cache_tree
        file_line_thread: "cache-tree.c:625:main"
        git_event_name: region_enter
        label: read
        msg: ~
        nesting: 2
      end:
        file_line_thread: "cache-tree.c:627:main"
        git_event_name: region_leave
        t_rel: 0.00003
  - ts: 4495
    pid: 1
    tid: 1
    name: unpack_trees
    cat: ""
    ph: X
    dur: 2443
    args:
      begin:
        category: unpack_trees
        file_line_thread: "unpack-trees.c:1724:main"
        git_event_name: region_enter
        label: unpack_trees
        msg: ~
        nesting: 1
      end:
        file_line_thread: "unpack-trees.c:1910:main"
        git_event_name: region_leave
        t_rel: 0.002455
  - ts: 4878
    pid: 1
    tid: 2
    name: traverse_trees
    cat: ""
    ph: X
    dur: 1568
    args:
      begin:
        category: unpack_trees
        file_line_thread: "unpack-trees.c:1817:main"
        git_event_name: region_enter
        label: traverse_trees
        msg: ~
        nesting: 2
      end:
        file_line_thread: "unpack-trees.c:1819:main"
        git_event_name: region_leave
        t_rel: 0.001568
  - ts: 6505
    pid: 1
    tid: 2
    name: Updating files
    cat: ""
    ph: X
    dur: 27
    args:
      begin:
        category: progress
        file_line_thread: "progress.c:268:main"
        git_event_name: region_enter
        label: Updating files
        msg: ~
        nesting: 2
      end:
        file_line_thread: "progress.c:336:main"
        git_event_name: region_leave
        t_rel: 0.000027
  - ts: 6549
    pid: 1
    tid: 2
    name: Filtering content
    cat: ""
    ph: X
    dur: 31
    args:
      begin:
        category: progress
        file_line_thread: "progress.c:268:main"
        git_event_name: region_enter
        label: Filtering content
        msg: ~
        nesting: 2
      end:
        file_line_thread: "progress.c:336:main"
        git_event_name: region_leave
        t_rel: 0.000031
  - ts: 7028
    pid: 1
    tid: 1
    name: prime_cache_tree
    cat: ""
    ph: X
    dur: 1005
    args:
      begin:
        category: cache-tree
        file_line_thread: "cache-tree.c:815:main"
        git_event_name: region_enter
        label: prime_cache_tree
        msg: ~
        nesting: 1
      end:
        file_line_thread: "cache-tree.c:822:main"
        git_event_name: region_leave
        t_rel: 0.001004
  - ts: 8050
    pid: 1
    tid: 1
    name: do_write_index
    cat: ""
    ph: X
    dur: 1234
    args:
      begin:
        category: index
        file_line_thread: "read-cache.c:3138:main"
        git_event_name: region_enter
        label: do_write_index
        msg: /Users/jsimms/focus/.git/index.lock
        nesting: 1
      end:
        file_line_thread: "read-cache.c:3141:main"
        git_event_name: region_leave
        t_rel: 0.001234
  - ts: 8050
    pid: 1
    tid: 2
    name: "data(index): write/cache_nr -> 94"
    cat: ""
    ph: X
    dur: 1226
    args:
      category: index
      file_line_thread: "read-cache.c:3102:main"
      git_event_name: data
      key: write/cache_nr
      nesting: 2
      t_abs: 0.009276
      t_rel: 0.001226
      value: "94"
  - ts: 8050
    pid: 1
    tid: 2
    name: "data(index): write/version -> 2"
    cat: ""
    ph: X
    dur: 1212
    args:
      category: index
      file_line_thread: "read-cache.c:3100:main"
      git_event_name: data
      key: write/version
      nesting: 2
      t_abs: 0.009262
      t_rel: 0.001212
      value: "2"
  - ts: 9096
    pid: 1
    tid: 2
    name: write
    cat: ""
    ph: X
    dur: 45
    args:
      begin:
        category: cache_tree
        file_line_thread: "cache-tree.c:529:main"
        git_event_name: region_enter
        label: write
        msg: ~
        nesting: 2
      end:
        file_line_thread: "cache-tree.c:531:main"
        git_event_name: region_leave
        t_rel: 0.000045
displayTimeUnit: ~
systemTraceEvents: ~
otherData: ~

